### PR TITLE
JSON Logging for Emulation Metrics

### PR DESCRIPTION
**Pull request by frodre**
This PR adds json metric logging for emulation diagnostic scripts that can be ingested by gcloud's logging infrastructure.  In addition to this a new regression CronJob (https://github.com/ai2cm/vcm-workflow-control/pull/127) will be run nightly to log important metrics from our training and prognostic runs nightly.

Metrics Added:
- training time
- prognostic run length
- MSE/bias scores on a test dataset for trained models

### CODE REVIEW

File: docker/artifacts/Dockerfile [hunk 1]
```diff
@@ -1,8 +1,25 @@
 FROM ubuntu:20.04
 
-RUN apt-get update && apt-get install -y python3 python3-pip
-COPY constraints.txt constraints.txt
-ENV PIP_CONSTRAINT=/constraints.txt
+RUN apt-get update && apt-get install -y python3 python3-pip curl gzip
+
+# Install argo for cronjob submission
+# https://github.com/argoproj/argo-workflows/releases/tag/v3.3.1
+RUN curl -sLO https://github.com/argoproj/argo-workflows/releases/download/v3.3.1/argo-linux-amd64.gz && \
```
Comment (frodre) [inline docker/artifacts/Dockerfile:7]: Adding Argo into this image adds ~100mb.  I'd understand if a separate image would be warranted here for the addition of Argo and some microphysics specific folders.

File: docker/fv3fit/Dockerfile [hunk 2]
```diff
@@ -35,6 +35,11 @@ RUN pip3 install --no-dependencies /fv3net/external/loaders
 COPY external/fv3fit /fv3net/external/fv3fit
 RUN pip3 install --no-dependencies /fv3net/external/fv3fit
 
+# Add emulation project scripts
+COPY projects/microphysics/scripts /fv3net/projects/microphysics/scripts
```
Comment (frodre) [inline docker/fv3fit/Dockerfile:39]: Was missing "score_training.py" script for the e2e workflow.

File: external/fv3fit/fv3fit/emulation/keras.py [hunk 3]
```diff
@@ -45,6 +45,6 @@ def score_model(model: tf.keras.Model, data: Mapping[str, tf.Tensor],) -> Scorin
         ``model``.
     """
     model = fv3fit.keras.adapters.ensure_dict_output(model)
-    prediction = model(data)
+    prediction = model.predict(data, batch_size=8192)
```
Comment (frodre) [inline external/fv3fit/fv3fit/emulation/keras.py:48]: The previous eager prediction step would OOM using 80,000 samples during scoring.  The `batch_size` of `.predict()` defaults to 32, so I chose a size that would run quickly and not use too much memory.

File: external/fv3fit/fv3fit/emulation/scoring.py [hunk 4]
```diff
@@ -4,18 +4,6 @@
 
 logger = logging.getLogger(__name__)
 
-SCALE_VALUES = {
```
Comment (frodre) [inline external/fv3fit/fv3fit/emulation/scoring.py:7]: Removed since it was no longer used and outputs an annoying error message.

File: external/fv3fit/fv3fit/emulation/scoring.py [hunk 5]
```diff
@@ -53,28 +41,27 @@ def score(target, prediction) -> ScoringOutput:
...
 
-    if rescale:
-        try:
-            target *= SCALE_VALUES[name]
-            prediction *= SCALE_VALUES[name]
-        except KeyError:
-            logger.error(f"No scaling value found for {name}. Leaving unscaled...")
+    # including for backwards compatibility for precip scores
+    if name == "total_precipitation":
+        precip_scaling = (1000 / (900 / (3600 * 24)),)  # mm / day
```
Comment (nbren12) [inline external/fv3fit/fv3fit/emulation/scoring.py:62]: Is the factor of 1000 correct? If the original units are kg/m2/s then the liquid water equivalent is 

1 [kg/m^2/s] / 1000 [kg/m^3] = 1  / 1000 [m / s] = 1 [mm/s] = 86400 [mm/d]

Nice to use variable names to name some of these convesion factors:

```
timestep = 900
seconds_in_day = 86400
density_liquid_water = 1000
mm_in_m = 1000
precip_scaling = mm_in_m / density_in_liquid_water * seconds_in_day / timestep
```

File: Makefile [hunk 6]
```diff
@@ -231,6 +232,15 @@ docker/fv3fit/requirements.txt:
 		external/loaders/setup.py \
 		external/vcm/setup.py
 
+docker/artifacts/requirements.txt:
+	cp constraints.txt $@
```
Comment (nbren12) [inline Makefile:236]: to allow overwriting the target if it exists without choking.
```suggestion
	cp -f constraints.txt $@
```

File: Makefile [hunk 7]
```diff
@@ -231,6 +232,15 @@ docker/fv3fit/requirements.txt:
...
+	# this will subset the needed dependencies from constraints.txt
+	# while preserving the versions
+	pip-compile --no-annotate \
+		--output-file docker/artifacts/requirements.txt \
```
Comment (nbren12) [inline Makefile:240]:
```suggestion
		--output-file $@ \
```

File: Makefile [hunk 8]
```diff
@@ -231,6 +232,15 @@ docker/fv3fit/requirements.txt:
...
+		external/artifacts/setup.py \
+		projects/microphysics/argo/requirements.in
```
Comment (nbren12) [inline Makefile:242]: hmm. maybe pull this requirements file someplace else. I sort of view the projects/microphysics folder as an event horizon. dependencees go in but not out.

File: docker/artifacts/Dockerfile (hunk 1 above)
Comment (nbren12) [reply docker/artifacts/Dockerfile:7]: I think this makes some sense IMO to make `artifacts` the image that encapsulates our submission/workflow management environment. Do we really not have the argo CLI version pinned anywhere else...pretty surprising that we haven't had any workflows break bc of it, but I guess argo's a well maintained/stable project.

File: docker/artifacts/Dockerfile [hunk 9]
```diff
@@ -1,8 +1,25 @@
...
+    gunzip argo-linux-amd64.gz && \
+    chmod +x argo-linux-amd64 && \
+    mv argo-linux-amd64 /usr/local/bin/argo && \
+    argo version
```
Comment (nbren12) [inline docker/artifacts/Dockerfile:11]: I love go btw...so nice that this works. We can make it a bit slicker using pipes

```suggestion
RUN curl -sL https://github.com/argoproj/argo-workflows/releases/download/v3.3.1/argo-linux-amd64.gz \
    | gunzip  > /usr/local/bin/argo \
    && chmod +x /usr/local/bin/argo \
    && echo
```

File: docker/artifacts/Dockerfile [hunk 10]
```diff
@@ -1,8 +1,25 @@
...
\ No newline at end of file
+RUN pip3 install --no-cache-dir --no-dependencies /tmp/artifacts
+
+COPY projects/microphysics/argo /projects/microphysics/argo
+COPY projects/microphysics/experiments /projects/microphysics/experiments
+COPY projects/microphysics/train /projects/microphysics/train
+COPY projects/microphysics/configs /projects/microphysics/configs
+
+ARG COMMIT_SHA_ARG
+ENV COMMIT_SHA=$COMMIT_SHA_ARG
```
Comment (nbren12) [inline docker/artifacts/Dockerfile:25]:
```suggestion
ENV COMMIT_SHA=$COMMIT_SHA_ARG

```

**Review by nbren12 — CHANGES_REQUESTED** [review]
Looking good!

I will request you undo the copy paste in the new metadata code since that will make it harder to switch to @AnnaKwa's new specification. For now, I would just drop that code in `fv3net.artifacts.metadata`, and open an issue to refactor.

File: workflows/diagnostics/fv3net/diagnostics/prognostic_run/emulation/single_run.py [hunk 11]
```diff
@@ -315,10 +332,14 @@ def upload_diagnostics_for_rundir(url: str, summary_only: bool, summary_name: st
...
 
+    summary_metrics = {"duration_seconds": get_duration_seconds(url)}
     for func in get_summary_functions():
         for key, val in func(ds):
             if re.match(summary_name, key):
-                log_summary(key, val)
+                summary_metrics[key] = val
+
+    log_summary_wandb(summary_metrics)
+    log_json_summary_gcloud(summary_metrics, tag)
```
[the remark this answers is not in the document]
Comment (frodre) [reply workflows/diagnostics/fv3net/diagnostics/prognostic_run/emulation/single_run.py:342]: Thanks for this info.  I switched back to printing the summary metrics as they are calculated.

File: projects/microphysics/scripts/score_training.py [hunk 12]
```diff
@@ -60,24 +61,43 @@ def main(config: TrainConfig, seed: int = 0, model_url: str = None):
...
+    gcloud_metrics = {
+        "json": dict(**summary_metrics),
+        "kind": "metrics",
+        "job_type": "train_score",
+        "url": config.out_url,
+    }
+    gcloud_metrics["logging.googleapis.com/labels"] = {
+        "commit_sha": os.getenv("COMMIT_SHA", ""),
+    }
+    print(json.dumps(gcloud_metrics))
```
[the remark this answers is not in the document]
Comment (AnnaKwa) [reply projects/microphysics/scripts/score_training.py:88]: FYI @frodre this is what I am putting in the step metadata: https://github.com/ai2cm/fv3net/compare/print-step-metadata?expand=1 (location will move to artifacts, but you can see what the dataclass attributes are)

File: external/fv3fit/fv3fit/emulation/scoring.py (hunk 5 above)
Comment (frodre) [reply external/fv3fit/fv3fit/emulation/scoring.py:62]: ```! rain1 has units m, and represents surface precip over dtp```

I'll update the scaling to indicate the units more clearly.

**Comment by nbren12** [comment]
@frodre I just noticed that circle ci uses argo 2.11: https://github.com/ai2cm/fv3net/blob/ff2130e971f05441413eda33b5cff52a3e5956f2/.circleci/config.yml#L157

Should we be consistent?

File: projects/microphysics/scripts/score_training.py [hunk 13]
```diff
@@ -48,36 +51,55 @@ def main(config: TrainConfig, seed: int = 0, model_url: str = None):
...
+    summary_metrics.update(
+        {f"score/test/{key}": value for key, value in test_scores.items()}
+    )
+
+    # Logging for google cloud
+    log_fact_json(data=summary_metrics)
+
     if config.use_wandb:
-        pred_sample = model.predict(test_set)
+        pred_sample = model.predict(test_set, batch_size=8192)
```
Comment (nbren12) [inline projects/microphysics/scripts/score_training.py:93]: Likewise nice to keep this hyper-parameter change in its own commit.

File: docker/artifacts/Dockerfile [hunk 14]
```diff
@@ -1,8 +1,24 @@
 FROM ubuntu:20.04
 
-RUN apt-get update && apt-get install -y python3 python3-pip
-COPY constraints.txt constraints.txt
-ENV PIP_CONSTRAINT=/constraints.txt
+RUN apt-get update && apt-get install -y python3 python3-pip curl gzip
+
+# Install argo for cronjob submission
+# https://github.com/argoproj/argo-workflows/releases/tag/v3.3.1
```
Comment (nbren12) [inline docker/artifacts/Dockerfile:6]:
```suggestion
# https://github.com/argoproj/argo-workflows/releases/tag/v2.11.6
```

**Review by nbren12 — APPROVED** [review]
LGTM. I'm going to request that you make sure the hyper parameter changes go into master as separate commits. Interactive rebase is fine, but also maybe easier to revert and open a micro PR with just those changes. (esp if the test failures below require more iteration).

File: projects/microphysics/scripts/score_training.py [hunk 15]
```diff
@@ -48,36 +51,55 @@ def main(config: TrainConfig, seed: int = 0, model_url: str = None):
...
+
     train_ds = config.open_dataset(
         config.train_url, config.nfiles, config.model_variables
     )
     test_ds = config.open_dataset(
         config.train_url, config.nfiles, config.model_variables
     )
 
-    train_set = next(iter(train_ds.unbatch().shuffle(100_000).batch(50_000)))
+    train_set = next(iter(train_ds.unbatch().shuffle(160_000).batch(80_000)))
```
Comment (nbren12) [inline projects/microphysics/scripts/score_training.py:75]: likewise, please isolate this into a separate commit or remove from the PR for now.

**Comment by nbren12** [comment]
One trick I do for making the history cleaner is to 1) manually remove the hyper parameter changes, 2) git commit, 3) git revert (this will become commit with just the parameter changes), 4) git rebase -i origin/master, and squash everything up to (3) into one commit, and reword the message of the revert commit

**Comment by frodre** [comment]
Thanks for the tip!

For this https://github.com/ai2cm/fv3net/pull/1789#discussion_r859233910 comment.  The change prevents the scoring script from OOM.  It doesn't change the calculated scores, only that it processes them sequentially before joining into an output.  I'm fine with micro PR to fix that for the sake of history, it'll just be broken temporarily on master probably.

**Comment by frodre** [comment]
Resolves #1792
